src/main/java/hudson/plugins/ec2/EC2Cloud.java

```diff
@@ -172,14 +174,18 @@
 
     private transient volatile AmazonEC2 connection;
 
-    protected EC2Cloud(String id, boolean useInstanceProfileForCredentials, String credentialsId, String privateKey,
-            String instanceCapStr, List<? extends SlaveTemplate> templates, String roleArn, String roleSessionName) {
+    protected EC2Cloud(String id, boolean useInstanceProfileForCredentials, String credentialsId, String privateKey, String sshKeysCredentialsId,
+                       String instanceCapStr, List<? extends SlaveTemplate> templates, String roleArn, String roleSessionName) {
         super(id);
         this.useInstanceProfileForCredentials = useInstanceProfileForCredentials;
         this.roleArn = roleArn;
         this.roleSessionName = roleSessionName;
         this.credentialsId = credentialsId;
-        this.privateKey = new EC2PrivateKey(privateKey);
+        this.sshKeysCredentialsId = sshKeysCredentialsId;
+
+        if (this.sshKeysCredentialsId == null && privateKey != null){
```

Comment:
This should probably be in `readResolve()`

Reply:
According to the comment at the readResolve() method call I interpret it as method caring about setting the template parents while this part checks if a migration to credentials is necessary here which in my opinion would not be part of this topic. Am I missing something here?